---
title: "HapMap Data"
description: |
     This page explains information on the data context and cleaning steps necessary to run GWAS and multiple hypothesis testing procedures in RStudio.
output: 
  distill::distill_article:
    toc: true
bibliography: Library.bib
---

In the [Data Analysis](https://multipletestingextension.netlify.app/analysisR.html) and [Hypothesis Testing in PLINK](https://multipletestingextension.netlify.app/HapMapPlink.html) tabs of this site, we will explain how to do a GWAS and determine a threshold for a set of genetic data. To follow along, download the `1_QC_GWAS.zip` file from [this page](https://github.com/MareesAT/GWA_tutorial/) [@mareesat]. This data comes from the International HapMap project (also known as “HapMap”). For more information on the data and project, check out [this tutorial](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/) [@hapmap].

## Data Loading and Organization

The following code chunks outline the steps of importing the genetic data.

If you have not installed the `snpStats` package, install it in the console by running the following code chunk. 

```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("snpStats")
```

Load libraries:

```{r}
library(snpStats)
library(tidyverse)
library(broom)
library(NatParksPalettes)
library(parallel)
library(GGally)     
```

Load the data, using the correct directory of where you put the `HapMap_3_r3_1.fam`, `HapMap_3_r3_1.bim`, and `HapMap_3_r3_1.bed` files from the `1_QC_GWAS.zip` folder. This process uses `read.plink()`, which reads a genotype matrix, information on the study’s individuals, and information on the SNPs.

```{r}
fam <- 'hapmapData/HapMap_3_r3_1.fam'
bim <- 'hapmapData/HapMap_3_r3_1.bim'
bed <- 'hapmapData/HapMap_3_r3_1.bed'

hapmap <- read.plink(bed, bim, fam)
```

## Understanding the Data

First, get information about the genotype data. We have 165 individuals and 1,457,897 SNPs.

```{r}
hapmap$genotypes
```

Next, look at the information we have on the individuals in the study. Theoretically, this gives information on family relationships with `pedigree`, `father`, and `mother`, but the `father` and `mother` variables contain only missing values. We also have information on the individual’s binary sex, with `1` representing male and `2` female. The `affected` column represents if the individual had the trait of interest or not, but there are many missing values in this column.

```{r}
head(hapmap$fam)
```

Finally, we can look at the information we have on each SNP. This tells us a few things:

- `chromosome` is the number chromosome (typically 1-23) that the SNP is located on. \
  - `1` is the largest chromosome (most SNPs) and chromosome size typically decreases from there.
- `snp.name` is the name of the SNP \
- `cM` stands for *centiMorgans*, which is a unit for genetic distance. It represents an estimate of how far SNPs are from one another along the genome. \
- `position` tells us the base pair position of the SNP, with position being being the first nucleotide in our DNA sequence. \
   - This number restarts from 1 at each chromosome. \
- `allele.1` is one of the alleles at this SNP, here the minor allele. \
- `allele.2` is the other allele at this SNP, here the major allele. \

```{r}
head(hapmap$map)
```

## Data Cleaning

One useful piece of information not contained in the data is the **minor allele frequency (MAF)**, which represents what proportion of people have the less commonly occurring base pair in the data. We can add this to our snpMatrix using the `snpstats` package and add MAF to `map`, our data frame that gives us SNP information.

```{r}
#calculate MAF
maf <- col.summary(hapmap$genotypes)$MAF

# add new MAF variable to map
map <- hapmap$map %>%
  mutate(MAF = maf)
head(map)
```

Just looking at the MAF for the first six SNPs in our data, we see that in some cases the minor allele frequency is 0. This means that the SNP is **monomorphic** - everyone in the dataset has the same genotype at these positions. We will remove these monomorphic SNPs - if everyone has the same alleles at a SNP, there is no variation and we cannot find an association between the minor allele and the trait.

It can also help to think about why we remove SNPs with a MAF of 0 in a mathematical way. If we are trying to fit a line between the trait of interest and SNP 1, we could model this in the following formats, with linear regression listed first and matrix notation second.

$$E[Y|\text{SNP1}] = \beta_0 + \beta1 \text{SNP1}$$
$$E[\bf{y}|\bf{X}] = \boldsymbol{\beta} X$$

Further exploring the matrix format, it would look like this:
$$X\boldsymbol{\beta} = \begin{bmatrix}
1 & 0 \\
1 & 0 \\
. & . \\
. & . \\
\end{bmatrix}
\begin{bmatrix}
\beta_0\\
\beta_1 \\
\end{bmatrix}$$

This problematic because we have linear dependence. You can get the column of minor allele counts by multiplying the intercept column by 0 - in other words, the minor allele count column is a linear combination of the intercept column. This makes our design matrix not be full rank, making $X^TX$ not invertible and the least squares estimator not defined.

Given all these reasons, we remove SNPs with a MAF of 0 using the code below.

```{r}
map <- map %>%
  filter(maf >0 )

dim(map)
```

After filtering, we have 1,283,751 SNPs remaining. Therefore, we removed 174,146 monomorphic SNPs.

Before moving on, we must complete one final data cleaning step. The `snpstats` package uses a format in which genotypes are coded as 01, 02, and 03, with 00 representing missing values. \

```{r}
hapmap$genotypes@.Data[1:5,1:5]
```

We will convert this to a 0, 1, and 2 format. Now the matrix represents the number of major alleles each person has at each SNP. \

```{r}
X <- as(hapmap$genotypes, "numeric")
X[1:5, 1:5]
```

Create `X.clean` by removing the monomorphic SNPs from `X`.

```{r}
map.clean <- map %>%
  filter(MAF >0)
X.clean <- X[,colnames(X) %in% map.clean$snp.name]
```


The data is now clean and ready to be used. Make sure you keep it loaded in your environment for future use. Check out the [Data Analysis](https://multipletestingextension.netlify.app/analysisR.html) tab next!

