---
title: "A final analysis of our HapMap data"
description: |
    How we combined our knowledge of multiple hypothesis testing and PLINK results to determine an appropriate significance threshold for our HapMap dataset
site: distill::distill_website
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Learn more about creating websites with Distill at:
# https://rstudio.github.io/distill/website.html
```

```{r}
library(tidyverse)
library(janitor)
library(snpStats)
library(NatParksPalettes)
```

## Read in data from plink

```{r}
adjust_results <- read.table("plinkData/as2.assoc.adjusted") %>%
  row_to_names(row_number = 1)

fam <- 'plinkData/HapMap_3_r3_1.fam'
bim <- 'plinkData/HapMap_3_r3_1.bim'
bed <- 'plinkData/HapMap_3_r3_1.bed'

hapmap <- read.plink(bed, bim, fam)
```

Combine snp information and p-value results

```{r}
adjust_with_position <- adjust_results %>%
  mutate(CHR = as.integer(CHR)) %>%
  left_join(hapmap$map %>%
              select(snp.name, position, chromosome), by = c("SNP" = "snp.name", "CHR" = "chromosome"))

#no scientific notation
adjust_with_position_no_sci <- adjust_with_position %>%
  filter(!str_detect(GC, "e-")) %>%
  mutate(GC = as.numeric(GC))

#deal with scientific notation
adjust_with_position_sci <- adjust_with_position %>%
  filter(str_detect(GC, "e-")) %>%
  separate(GC, into = c("decimal", "digits"), sep = "e-0", remove =FALSE) %>%
  mutate(decimal = as.numeric(decimal), 
         digits =as.numeric(digits), 
         GC = decimal * 10^(-digits)) %>%
  select(-decimal, -digits)

adjust_with_position <- bind_rows(adjust_with_position_sci, adjust_with_position_no_sci)
```

## Analysis

Most significant SNPs

```{r}
adjust_with_position %>%
  arrange(GC) %>%
  head(10)
```


Create manhattan plot

```{r, cache=TRUE}
adjust_with_position %>%
  mutate(minuslogp = -log10(GC),
         CHR = as.factor(CHR)) %>%
  ggplot(aes(x = CHR, y = minuslogp, group = interaction(CHR, position), color = CHR)) + 
  geom_point(position = position_dodge(0.8)) + 
  labs(x = 'chromosome', y = expression(paste('-log'[10],'(p-value)')))+
  theme_classic()+
  scale_color_manual(values=natparks.pals("DeathValley",24))+
  theme(legend.position = "none")
```


```{r}
hapmap$fam %>%
  count(affected)
```

Account for correlation

```{r}
chr1_200 <- hapmap$genotypes[1:165, 1:200]
hapmap.ld <- ld(chr1_200, depth = 199, stats = "R.squared", symmetric = TRUE)
color.pal <- natparks.pals("Acadia", 10)
image(hapmap.ld, lwd = 0, cuts = 9, col.regions = color.pal, colorkey = TRUE)
```

```{r}
#get monomorphic SNPs only
maf_chr1_200 <- col.summary(chr1_200)$MAF
mono <- which(maf_chr1_200 == 0)

# calculate LD on polymorphic SNPs only
hapmap.ld.nomono <- ld(chr1_200[,-mono], depth = 199-length(mono), stats = "R.squared", symmetric = TRUE)

# plot 
image(hapmap.ld.nomono, lwd = 0, cuts = 9, col.regions = color.pal, colorkey = TRUE)
```

