---
title: "A final analysis of our HapMap data"
description: |
    How we combined our knowledge of multiple hypothesis testing and PLINK results to determine an appropriate significance threshold for our HapMap dataset
site: distill::distill_website
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Learn more about creating websites with Distill at:
# https://rstudio.github.io/distill/website.html
```

```{r}
library(tidyverse)
library(janitor)
library(snpStats)
library(NatParksPalettes)
```

### Load hapmap data

Create a null trait and write a file to the plink data folder.

```{r}
fam <- 'plinkData/HapMap_3_r3_1.fam'
bim <- 'plinkData/HapMap_3_r3_1.bim'
bed <- 'plinkData/HapMap_3_r3_1.bed'

hapmap <- read.plink(bed, bim, fam)
```

```{r}
manhattan_data <- cbind(hapmap$fam %>% select(1:2), trait = rnorm(n = 165, mean = 0, sd = 1))
write_delim(manhattan_data, "plinkData/manhattandata")
```

Use command to run GWAS in PLINK, accounting for correlated SNPs
- ./plink --bfile HapMap_3_r3_1 --assoc --adjust --pheno manhattandata --out as2

Read in results from PLINK and create Manhattan plot

```{r}
results <- read_table("plinkData/as2.qassoc.adjusted")

results_with_position <- results %>%
  mutate(CHR = as.integer(CHR)) %>%
  left_join(hapmap$map %>%
              select(snp.name, position, chromosome), by = c("SNP" = "snp.name", "CHR" = "chromosome"))
```

82146
```{r, cache=TRUE}
results_with_position %>%
  mutate(minuslogp = -log10(GC),
         CHR = as.factor(CHR)) %>%
  ggplot(aes(x = CHR, y = minuslogp, group = interaction(CHR, position), color = CHR)) + 
  geom_point(position = position_dodge(0.8)) + 
  labs(x = 'chromosome', y = expression(paste('-log'[10],'(p-value)')))+
  theme_classic()+
  geom_hline(yintercept = 7.135789, color = "navy", linetype = "dashed")+
  geom_hline(yintercept = 1.30103, color = "navy", linetype = "dashed")+
  scale_color_manual(values=natparks.pals("DeathValley",24))+
  theme(legend.position = "none")
```

## Complete replications

Simulation based approach. Approximately 14.45 minutes to run GWAS in plink, then 3 minutes 20 seconds to load files into R. 

```{r, eval=FALSE}
create_quantitative_trait <- function(i){
  y <- rnorm(n = 165, mean = 0, sd = 1) 
}

traits <- as.data.frame(replicate(1000, create_quantitative_trait()))

traits_identified <- cbind(hapmap$fam %>%
        select(1:2), traits)

write_delim(traits_identified, "plinkData/traits_identified")
```

./plink --bfile HapMap_3_r3_1 --assoc --pheno traits_identified --all-pheno --pfilter 1e-3

```{r, cache=TRUE}
dataFiles <- lapply(Sys.glob("plinkData/plink.P*.qassoc"), read_table)

pvalues <- sapply(dataFiles, function(x) min(x$P, na.rm=TRUE))

as.data.frame(pvalues) %>%
  ggplot(aes(x=pvalues))+
  geom_density(fill = "cadetblue")+
  theme_classic()+
  annotate(geom = "text", color = "red", x = 1e-05, y = 250000, label = "0.05 quantile:\n7.31495e-08", family = "mono", cex = 3)+
  geom_vline(xintercept = 7.31495e-08, color = "red", linetype = "dashed")+
  labs(x="P-values", y = "Density", title = "Distribution of minimum p-values for 1000 replications")+
  geom_curve(aes(x = 1e-05, y = 270000, xend = 1e-06, yend = 350000), 
             arrow = arrow(length = unit(0.03, "npc")), curvature = 0.3, color = "red")+
  theme(plot.title.position = "plot", 
        plot.title = element_text(family = "mono"), 
        axis.title = element_text(family = "mono"), 
        axis.text = element_text(family = "mono"))

quantile(pvalues, 0.05)
```

