---
title: "Multiple Hypothesis Testing"
description: "An explanation of the complexities to making statistically significant conclusions when working with genetic data."
output: 
  distill::distill_article:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# Learn more about creating websites with Distill at:
# https://rstudio.github.io/distill/website.html
```

## Genome Wide Association Studies

To understand the need for multiple hypothesis testing, we need to be familiar with Genome Wide Association Studies(GWAS) and the complexities to working with genetic data.

### Explanation of GWAS

If we compare any two human genomes, they are nearly identical. However, places where DNA sequences differ are known as genetic variants. Once there is a collection of variants to be studied, GWAS attempts to determine which genetic variants are associated with an illness (Type One Diabetes, Crohn's Disease) or trait of interest (height, blood pressure).

There are multiple types of genetic variants, but GWAS is focused on variants that are exchanges of one nucleotide base pair for another, otherwise known as single nucleotide polymorphisms(SNPs). Figure 1 is a visual example of a SNP.

<strong>SNPs</strong> : DNA variations that occur when a single nucleotide base pair in the genome sequence is altered (AT vs GC at a single position).

<center>
![Figure 1: SNP Example](images/geneticvariants.png){width=40%}
</center>

### Working with Genetic Data

When working with genetic data and running GWAS, the number of SNPs to be studied is commonly greater than 1,000,000. Because of this, we cannot simply fit a multiple linear regression model to predict the existence of an illness or the value of a trait of interest. As an equation, this would be equivalent to <strong>E[y | x] = Íûµ + Íûµ1x + ... + Íûµnx</strong>, where y is the trait and x is the number of minor alleles.

In this situation, our goal would be to find a line of best fit where ‚Äúbest‚Äù is described as the minimized sum of squared residuals, which requires complicated derivations. This would be quite time consuming and difficult to find the optimal Íûµ0 and Íûµ1...Íûµn considering how large the number of predictors is. Even if we used a different technique such as matrix derivation, because there are more SNPs(columns) than individuals we are studying(rows), our matrix is not invertible and thus our least squares estimator is not defined. This conclusion is proven through linear algebra.

Instead, we utilize <strong>Marginal Regression</strong>. As an equation, this would be equivalent to <strong>E[y | xj] = Íûµ + Íûµ1xj + ... + Íûµnxj</strong>, where y is the trait and x is the number of minor alleles at that SNP. From these marginal regressions, we receive test statistics and p-values for each SNP. Once we have a p-value for each SNP, we can begin to visualize these values and make decisions of whether or not a SNP is associated with our trait of interest. Figure 2 is an example of how we can visualize p-values, known as a Manhattan Plot. The y-axis is -log10 transformed so that smaller p-values appear as higher on the plot. For now, ignore the horizontal blue line.

<center>
![Figure 2: Manhattan Plot](images/manhatplot.png){width=90%}
</center>

## Hypothesis Testing in GWAS

In order to make these decisions of whether or not an association exists, we utilize hypothesis testing. As you read this section, note that the null hypothesis in GWAS is that there is no relationship between a SNP and the trait of interest. 

### Hypothesis Testing Review

To review, hypothesis testing is predicated on the comparison of a p-value and a predetermined significance threshold, commonly represented as ùõº and equal to 0.05. There are two hypotheses, the null hypothesis and the alternative hypothesis. If a p-value is less than the significance threshold, we reject the null hypothesis and therefore conclude the alternative hypothesis to be true. In context, we would conclude that there is a relationship between that SNP and the trait of interest. On the other hand, if a p-value is greater than the significance threshold, we cannot reject the null hypothesis and therefore conclude the null hypothesis to be true. In context, we would conclude that there is not a relationship between that SNP and the trait of interest. Figure 3 is a table that summarizes the decisions we can make and the consequences of those decisions. We desire to correctly reject the null as much as possible(power) while avoiding both falsely rejecting the null(Type 1 Error) and failing to reject the null when we should (Type 2 Error).

<center>
![Figure 3: Hypothesis Testing Decisions](images/table.png){width=90%}
</center>

### The Multiple Testing Problem

Aside from the added GWAS context, our explanation should be identical to what you would encounter in an introductory statistics course. However, there is additional complexity in GWAS because of the aforementioned need for marginal regression. Recall, we are not running one hypothesis test, but millions, and this drastically effects our ability to conclude that a SNP is significantly associated. 

To illustrate, imagine there is a bucket with ten balls inside, nine are red and one is blue, as seen in Figure 4. If you reach in the bucket once, your chance of grabbing a red ball is 90% (9/10). If you reach in the bucket five times, independently, your chance of grabbing a red ball all five times is 59% (.9^5). In this five-ball scenario, your chance of grabbing the blue ball <strong>at least once</strong> increases from 10%(one-ball scenario) to 41%. As expected, if you reach in the bucket ten times, independently, your chance of grabbing a red ball all ten times is even smaller, 35% (.9^10), and your chance of grabbing the blue ball <strong>at least once</strong> increases to 65%. Overall, the more times you reach in the bucket, the higher the chance of grabbing at least one blue ball.

<center>
![Figure 4: Bucket Example](images/bucket.png){width=50%}
</center>

With this in mind, consider every time you reach into the bucket as a hypothesis test and the event of grabbing the blue ball as Type 1 Error. As the number of hypothesis tests increases, the chance of at least one Type 1 Error increases. In words, if we reject the null hypothesis when we witness a "rare" event, it will be easier to find rare events and furthermore easier to mistakenly think that an event is truly rare as the number of tests get increasingly larger. Clearly, this is a major problem with the multiple hypothesis testing procedure that is needed to make decisions in GWAS.

This concept is further supported by this equation, <strong>P(at least one Type 1 Error) = 1 - (1 - ùõº)^n</strong>, which holds due to a number of proofs that we will point out but won't fully explain at this time. First, according to mathematical statistics proofs, ùõº is equal to the chance of Type 1 Error for a singular hypothesis test. Second, according to probability proofs, P(A >= 1) = 1 - P(complement of A) where A = Type 1 Error. Lastly, we can raise our equation to the power of n, with n being the number of hypothesis tests we are running, because we assume each test to be independent. 

### Determining a Threshold

This probability of at least one Type 1 Error is better known as the <strong>Family Wise Error Rate (FWER)</strong>. Rather than ùõº = 0.05, we often want FWER = 0.05. Looking back at the equation for FWER, we know the number of tests to be conducted will be over a million, so we must adjust the significance threshold.

To do this, there is a widely used approach known as the Bonferroni Correction. This correction is very straightforward, simply dividing the desired FWER by the number of hypothesis tests conducted. Here is the Bonferroni Correction as an equation. <strong>ùõº = FWER / Number of Hypothesis Tests</strong>. For example, if we conduct 1,000,000 tests and would like a FWER of 0.05, ùõº must equal 5 x 10 ^ -8. With this new threshold in mind, let's look back at a Manhattan Plot of SNPs that are truly null. If these SNPs are truly null, any point above the horizontal threshold line is considered a Type 1 Error. Figure 5 has two horizontal lines, one at ùõº = 0.05 and one at ùõº = 0.00000005. Imagine the proportion of Type 1 Error if our threshold was not adjusted.

<center>
![Figure 5: Manhattan Plot With Thresholds](images/finalmanhattan.png){width=90%}
</center>

## False Discovery Rate

While the Bonferroni Correction is straightforward, easy to use, and clearly effective (as shown by Figure 5), there are limitations and shortcomings to this procedure. When hypothesis tests are correlated, which they commonly are in GWAS, the Bonferroni Correction is overly conservative. This means that the adjusted significance threshold is too small, so we will not be able to reject the null as much, limiting power. Without diving into how and why SNPs are correlated, this truth about Bonferroni can be understood by analyzing the equation. When SNPs are correlated, there are essentially less SNPs being studied, and because we divide by the number of hypothesis tests conducted, this number in the denominator is larger than it should be, causing ùõº to be smaller than it should be.

To account for correlation, there are simulation-based approaches to determining a threshold. For more about this, please continue to read through this website.

Besides simulation-based approaches, there are also alternatives to the FWER. One of these alternatives is known as the <strong>False Discovery Rate (FDR)</strong>. FDR is defined as the proportion of false rejections of the null among all significant results. A FDR of 5% means that, among all features called significant, 5% of these are truly null. Essentially, the FWER controls the probability of all Type 1 Error, while the FDR allows more Type 1 Error but controls how many there are in proportion to true rejections of the null. While FDR may allow for a higher Type 1 Error rate, the trade-off is higher power. Furthermore, this power advantage increases with an increasing number of hypothesis tests.

For more on the FDR compared to the FWER, check out this journal. <a href="https://www.stat.purdue.edu/~doerge/BIOINFORM.D/FALL06/Benjamini%20and%20Y%20FDR.pdf">FDR Journal</a>

