---
title: "Run GWAS 1000 times without PLINK"
author: "Erin Franke"
date: "2022-11-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(janitor)
library(snpStats)
library(NatParksPalettes)
library(broom)
```

### Load hapmap data

Create a null trait and write a file to the plink data folder.

```{r}
fam <- 'plinkData/HapMap_3_r3_1.fam'
bim <- 'plinkData/HapMap_3_r3_1.bim'
bed <- 'plinkData/HapMap_3_r3_1.bed'

hapmap <- read.plink(bed, bim, fam)
```

```{r}
hapmap.geno <- as(hapmap$genotypes, "numeric")
dim(hapmap.geno)
```

```{r}
# calculate MAF
maf <- col.summary(hapmap$genotypes)$MAF

# find monomorphic SNPs
monomorphic <- which(maf == 0) 

# filter genotype matrix to remove monomorphic SNPs
hapmap.geno <- hapmap.geno[,-monomorphic]

# check the dimensions after filtering
dim(hapmap.geno)
```

```{r}
j = 1
do_one_rep_hapmap <- function(){
  print(j)
  j <<- j+1
  # simulate null trait
  y <- rnorm(n = 165, mean = 0, sd = 1) 
  # implement GWAS
  pvals <- c()
  for(i in 1:1283751){
    mod <- lm(y ~ hapmap.geno[,i])
    pvals[i] <- tidy(mod)$p.value[2]
  }
  # record the smallest p-value
  min(pvals)
}

# then repeat many times
set.seed(494) 
hapmap.reps <- replicate(1000, do_one_rep_hapmap())

# then use the quantile function to find the lower 5th percentile
quantile(hapmap.reps, probs = c(0.05))
```

